drivers: watchdog: document and enforce semantics of `WDT_DISABLE_AT_BOOT`

mathieuchopstm opened the pull request:
Fixes #86223 (hopefully).

The `WDT_DISABLE_AT_BOOT` Kconfig option has been interpreted as "enable the WDT if disabled" by *some* driver implementers, even though this is at best a hack. Document its true purpose explicitly and update all drivers to follow the now-documented semantics.

I took a conservative approach and only updated drivers which were explicitly playing the `if (!IS_ENABLED(WDT_DISABLE_AT_BOOT)) { enable_myself(); }` game, but maybe other drivers need to be updated too.

mathieuchopstm committed:
drivers: watchdog: document semantics of WDT_DISABLE_AT_BOOT

WDT_DISABLE_AT_BOOT should only be used to forcefully disable watchdog
timers that are enabled by default after system reset. Update the Kconfig
help text of WDT_DISABLE_AT_BOOT and its dependency HAS_WDT_DISABLE_AT_BOOT
to make this clear.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: watchdog: align drivers to (HAS_)WDT_DISABLE_AT_BOOT semantics

There are many drivers which control a disabled-by-default watchdog timer
but take the liberty of selecting HAS_WDT_DISABLE_AT_BOOT and interpreting
WDT_DISABLE_AT_BOOT=n as "enable the timer", which does not correspond to
the semantics of this option.

Update all such drivers to no longer select HAS_WDT_DISABLE_AT_BOOT and
ignore the WDT_DISABLE_AT_BOOT option.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

etienne-lms commented on drivers/watchdog/Kconfig:
```diff
@@ -13,12 +13,22 @@ if WATCHDOG
 	depends on HAS_WDT_DISABLE_AT_BOOT
 	help
-	  Disable watchdog at Zephyr system startup.
+	  Disable watchdog timers during Zephyr system startup.
```
Disable watchdog timers *"during Zephyr system startup"* or *"at Zephyr system startup"*?
*"during"* could mean disable while system is starting up then enable them back.

etienne-lms commented on drivers/watchdog/Kconfig:
```diff
@@ -13,12 +13,27 @@ if WATCHDOG
+
+	  Note that enabling this option may render certain watchdog timers
+	  unusable by the application, depending on the hardware implementation
+	  (e.g., if the timer cannot be re-enabled after being disabled).
```
In case a HW watchdog is not capable of being disabled, should the driver fail to build when `CONFIG_WDT_DISABLE_AT_BOOT` is enabled?

yvesll commented:
> I'm also wondering if `WDT_DISABLE_AT_BOOT` shouldn't be changed to `default y` since that is kinda what seems to be the _expected behavior_ (i.e., watchdog would be opt-in rather than opt-out).

https://github.com/zephyrproject-rtos/hal_nxp/blob/c4824f704c4fded29e7a37db5019e2744560bf26/mcux/mcux-sdk-ng/devices/MCX/MCXE/MCXE31B/system_MCXE31B.h#L54
In hal_nxp, some devices require the macro.

yvesll commented:
> There is **definitely** a need for "set some Kconfig and get a watchdog setup during boot" - that's what `WDT_DISABLED_AT_BOOT=n` was abused for 🙂 - but I'm not super inspired about _how to implement that_, which is kinda why I'm trying to get attention / feedback on this topic.

Totally agree with this, current `WDT_DISABLED_AT_BOOT` did not have a consistent behavior on watchdog drivers and it mostly like to remind developers be careful with watchdog operation. Maybe you can open a RFC for this topic?

mathieuchopstm committed:
doc: migration: 4.4: add entry about CONFIG_WDT_DISABLE_AT_BOOT

The unclear semantics of CONFIG_WDT_DISABLE_AT_BOOT have now been set in
stone, but this change will affect users which relied on a specific
interpretation of CONFIG_WDT_DISABLE_AT_BOOT which doesn't match the
semantics that have been agreed upon.

Document this change in the migration guide, indicating that watchdogs must
now be configured explicitly by the applications if required.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

etienne-lms reviewed (commented):
Comment on use of "during" wording.

mathieuchopstm commented on drivers/watchdog/Kconfig:
```diff
@@ -13,12 +13,27 @@ if WATCHDOG
+	  HAS_WDT_DISABLE_AT_BOOT must ignore this option.
+
+	  When enabled, all watchdog timer instances enabled in Devicetree must
+	  be disabled by their respective drivers during instance initialization.
```
I don't think there's any ambiguity here about `during` meaning `as part of` here, especially given the option's name is `AT_BOOT`

kevinwang821020 reviewed (approved):
The part of AndesTech Platforms is LGTM.
